Step3 다시 구현하였습니다.

ehgud0670 opened the pull request:
* 기존의 Tag 를 쓰던 방식을 버리고 상속과 프로토콜을 이용하는 구조로 리팩토링하였습니다.  
  * 음료 버튼과 음료 레이블은 상속과 프로토콜을 이용해 각자 다른 음료를 리턴하도록 하였습니다. 
    * 음료를 UILabel, UIButton 할것없이 리턴하는 역할이므로 프로토콜 `BeverageGeneratable`을 채택하면 음료를 리턴하게끔 하여 특정 클래스에 구애받지 않도록 하였습니다. 
  * 가격 버튼은 상속만을 이용해 각자 다른 돈을 핸들러에 던져주도록 하였습니다. 

* 이전 step3 PR에는 빌더를 썼는데 이를 제거하였습니다.

* 이전 step3 PR에는 Kind라는 것을 썼는데 제거하고 음료객체가 Hashable을 채택하여 인스턴스 객체끼리 비교가능하고,  딕셔너리의 키로도 사용할 수 있게끔 리팩토링하였습니다.

ehgud0670 committed:
Refactor: package 속성을 enum에서 string으로 바꿈
	이유: enum 타입으로 선언하는 속성은 잘 변경하지 않는 경우에만 하는게 좋다.
	      왜냐하면 enum 타입으로 선언한 경우, 변경에 유연하지 않고 경직된 경우가 많다.
	      enum에 새로운 case를 추가할 때 고쳐야 하는 것이 여러 곳이라면 나쁜 냄새가 난다고 생각하자

ehgud0670 committed:
Refactor: 전체 상품 재고를 (딕셔너리로 표현하는) 제품별로 표현하는 메소드

ehgud0670 committed:
Feat: 재고의 모든 음료를 탐색하는 메소드 추가

ehgud0670 committed:
Refactor: 살수있는 음료들 딕셔너리로 리턴되게끔 메소드 수정함

ehgud0670 committed:
Refactor: forEach 문 사용하게끔 코드 수정함

ehgud0670 committed:
Style: 파일 위치 옮김

ehgud0670 committed:
Refactor: Calculable 도입 및 Cashier에 적용
	이유: 자판기 앱과 Cashier간에 의존성을 프로토콜로 느슨하게 하기 위해서
	      즉, 의존성 역전 원칙을 사용하려고 적용하였다.

ehgud0670 committed:
Style: 딕셔너리를 리턴하는 메소드만 extension으로 따로 분리함

ehgud0670 committed:
Refactor: sell 메소드의 리턴타입을 Result<Beverage,SellError>로 바꿈
	이유 : sell메소드를 사용할때 실패한 경우,
	       무엇때문에 실패했는지 알리고 싶어서

ehgud0670 committed:
Style: 생성자 개행 정리함

ehgud0670 committed:
Feat: 딕셔너리의 key로 사용할 Kind 구조체 구현 및 속성으로 추가
Deleted: 기존에 사용하지 않는 Hashable을 채택한 Beverage를 제거함

* 나중에 딕셔너리의 key로 사용하기 위해 internal 접근 제한자로 둠

* Kind를 구현한 궁긍적인(Ultimate) 이유

 기존의 키는 타입으로 구분하였다.
 하지만 타입으로 음료를 구분할 수 없다.
 왜냐하면 볼륨(volume)에 따라 엄연히 다른 음료라서 타입만으로는 구별을 할 수 없기 때문이다.
 예를 들어 칸타타 프리미엄 라떼 175ml와 프리미엄 라떼 275ml는 타입은 같지만 volume이 다르므로 엄연히 다른 음료이고
 딕셔너리에서도 다른 음료로서 구분되어야 한다.
 따라서 Kind 구조체를 생성해 이름(name)과 볼륨(volume)으로 음료를 구분하게끔 한 것이다.

ehgud0670 committed:
Refactor: 리턴값인 딕셔너리의 key를 Kind 구조체로 수정하고, value를 Int로 수정함

ehgud0670 committed:
Refactor: Kind 구조체의 속성으로 brand와 price도 추가
	이유: 캡슐화 하기 위해서 추가함

ehgud0670 committed:
Refactor: 1. 캡슐화되도록 빌더 패턴 구현 및 적용함
	  2. 빌더를 클래스로 선언하여 상속을 하도록 함
	  3. 빌더에 정적 속성을 생성하여 재사용하도록 함

ehgud0670 committed:
Style: 개행 정리함

ehgud0670 committed:
Feat: 뷰컨트롤러에 자판기 객체 추가
	* 자판기 객체는 음료를 6개 가지고 있다.

	 1. PrimiumLatte175ml
	 2. PrimiumLatte275ml
	 3. DietCola250ml
	 4. DietCola591ml
	 5. CookieCream235ml
	 6. CookieCream350ml

ehgud0670 committed:
Refactor : 상속이 필요없는 클래스들을 final로 제한함

ehgud0670 committed:
Feat: 앞으로 사용할 커스텀 뷰들 생성함
	* 상속이 필요없으므로 final로 선언함

ehgud0670 committed:
Refactor: 각 커스텀 뷰에 필요한 생성자들 구현함

ehgud0670 committed:
Refactor: 관련된 뷰들 하나의 파일로 통합함

ehgud0670 committed:
Feat: tag 관련 속성 추가

ehgud0670 committed:
Refactor: 두 커스텀 뷰가 BeverageTagControl를 채택하도록 코드 수정함

ehgud0670 committed:
Refactor: 프로토콜 BalanceTagControl을 준수하도록 구현함

ehgud0670 committed:
Feat: 음료 이미지들 추가

ehgud0670 committed:
Feat: 스토리보드에 자판기 음료들 뷰 추가
	* 태그도 설정함

ehgud0670 committed:
Deleted: 자판기의 초기 재고들 삭제함

ehgud0670 committed:
Feat: 스토리보드의 뷰들 뷰컨트롤러에 선언함
    * 레이블은 UIOutlet Collection 사용함 => 모든 레이블 추가
    * 버튼 액션 메소드만 구현함 => 버튼 액션에 모든 버튼 추가함

ehgud0670 committed:
Refactor: 1.음료 객체의 제조일자를 매개변수에서 주입받지 않고 실시간 Date객체를 생성하도록 코드 수정함
	  2.우유 객체의 유통기한을 제조일자의 일주일 뒤로 설정함 => 이를 위해 제조일자 속성 접근을 private에서 internal로 변경함
	  3. 모든 음료 클래스의 생성자에서 옵셔널 조건을 지움 init?() => init()

ehgud0670 committed:
Refactor: 리턴값이 음료객체가 되도록 코드 수정함

ehgud0670 committed:
Refactor: 타입 변경함
	이유: 이전 버전이 타입을 실수로 기입한 것이라서

ehgud0670 committed:
Refactor: '추가' 버튼을 누르면 해당 음료가 자판기 구조체에 추가되도록 구현함

ehgud0670 committed:
Refactor: 바뀐 코드에 맞게 테스트 코드가 통과되도록 수정함

ehgud0670 committed:
Refactor: 음료 배열인 stock이 변하면 notification post 하도록 코드 변경함

ehgud0670 committed:
Refactor: balance의 값이 변하면 Notification post하도록 코드 수정함

ehgud0670 committed:
Feat: 옵져버 패턴을 이용해 추가 '버튼'을 누르면 레이블이 업데이트 되도록 기능 구현

ehgud0670 committed:
Refactor: 자판기를 구조체에서 클래스로 변경 및 적용
	이유 : Notification으로 객체를 주고받을때
 		자판기 객체의 주소값을 이용하려고

ehgud0670 committed:
Refactor: notification 필요없으므로 _로 표시하고
	  gurad 문 제거하고 자판기 객체를 바로 사용하도록 코드 수정함

	  이유: addObserver에서 object에 자판기 객체를 등록했으므로
		동작할때 같은 인스턴스임이 자명하다.
	        따라서 notification 매개변수 없이 바로
		자판기 객체 속성을 사용하도록 하였다.

ehgud0670 committed:
Refactor : 메소드 만들어서 레이블 객체 안에서 레이블 텍스트를 변경하도록 함

ehgud0670 committed:
Refactor: 의미없는 @objc 제거함

ehgud0670 committed:
Feat: 현재 잔액을 보여주는 커스텀 레이블 구현함

ehgud0670 committed:
Feat: 금액 관련 뷰들 스토리보드에 추가함
Refactor: 추가 버튼의 identifier를 PlusBeverageButton으로 정정함
	  이유: 이전에 identifier를 PlusBalanceButton으로 잘못 할당한거라
	         이번에 수정함

ehgud0670 committed:
Refactor: 텍스트에서 '잔액 : ' 을 반영하도록 코드 수정함

ehgud0670 committed:
Feat: 뷰컨트롤러에 금액 관련 스토리보드 객체들 선언함

ehgud0670 committed:
Refactor: 금액 추가 버튼을 통해 금액을 받으면 자판기 객체에 돈이 추가되도록 구현함

ehgud0670 committed:
Style: 관련있는 것끼리 모이도록 코드 옮김

ehgud0670 committed:
Refactor: Cashier가 변하면 자판기 객체가 NotificationCenter.default를
	  통해 보내지도록 코드 수정함
	이유: 뷰컨트롤러에서 관찰할수 있는 건 자판기 객체뿐이라서 코드 수정하였다.

ehgud0670 committed:
Feat: NotificationCenter의 addObserver을 통해 금액 추가 버튼을 누르면
	현재 금액이 반영되도록 코드 구현함

ehgud0670 committed:
Refactor: 처음에 잔액 레이블이 자판기 객체의 현재 잔액을 반영하도록 코드 수정함

ehgud0670 committed:
Style: 메소드의 의미를 더 반영하도록 함수명 바꿈

ehgud0670 committed:
Docs: 결과 화면 추가

ehgud0670 committed:
Style: 선언부분 사이에 개행 정리함

ehgud0670 committed:
Refactor: beverageItemByTag 속성 추가 및 protocol 확장하여 beverage() 메소드를 구현함
 	* BeverageItemByTag 객체를 내부 속성으로 소유하는 이유
	  : 재사용하기 위해서

	* beverage() 메소드를 protocol를 확장하여 구현한 이유
	  : 중복되는 메소드라서 프로토콜을 확장하여 구현하였다.

ehgud0670 committed:
[#step3] refactor: 상속 안하는 클래스들 모두 final 키워드를 추가

ehgud0670 committed:
[#step3] chore: 클래스 선언부 부분 줄간격 삭제

ehgud0670 committed:
[#step3] chore: 1x 사이즈 이미지들 삭제

ehgud0670 committed:
[#step3] refactor: 전반적으로 오토레이아웃 적용함

ehgud0670 committed:
[#step3] chore: 한 클래스당 한 파일로 분리 및 클래스이름에 Plus 삭제

	이유: 하나의 파일당 하나의 클래스만 있어야만 분별하기 쉽기때문에 파일을 분리했고,
	      Plus 라는 이름이 굳이 클래스에 없어도 이해가 되는 것 같아 Plus를 삭제함

ehgud0670 committed:
[step] refactor: 노티 네임들을 보내는 곳으로 옮기고, 옵저버를 등록하는 메소드만 분리

ehgud0670 committed:
[#step3] chore: 설정하는 메소드의 이름을 바꾸고, 레이블 텍스트의 기본값 설정

ehgud0670 committed:
[#step3] feat: corner가 둥근 뷰를 만들어서 적용

ehgud0670 committed:
[#step3] feat: 버튼의 이미지가 커져지도록 기능 구현

ehgud0670 committed:
[#step3] feat: 잔고 버튼 tag 관련 코드 삭제하고 상속을 적용함

ehgud0670 committed:
[#step3] feat: VC가 deinit 될때 등록된 옵저버를 삭제하는 기능 구현

ehgud0670 committed:
[#step3] refactor: token을 발행하여 제대로 옵저버를 삭제하도록 기능 구현

ehgud0670 committed:
[#step3] style: 해당 객체들은 observer 역할을 하므로 옵저버라고 이름 수정

ehgud0670 committed:
[#step3] refactor: 상수에서 enum case 로 교체

	이유: OOP 관점으로도 enum을 사용해도 될 것 같아서 리팩토링한다.

ehgud0670 committed:
[#step3] refactor: 빌더 패턴 삭제

	이유: 지금은 빌더가 복잡도만 더할 뿐이다.
	불필요하므로 삭제한다.
	하지만 나중에 다시 기준을 세워서 만들어야 한다면 다시 만들도록 하자

ehgud0670 committed:
[#step3] refactor: tag로 작동하는 방식을 버리고 상속을 통한 다형성을 적용

ehgud0670 committed:
[#step3] refactor: 음료를 생성하는 protocol을 만들어 Label, Button 객체에 모두 적용
		이유: 음료 객체를 반환하는 작업은 특정 객체에 갇혀 있으면 안되고 음료 객체를 반환할 책임이 있는 객체들 모두에 적용 가능 해야 한다.
		      따라서 프로토콜을 만들어 채택하게 하고 필요없는 상속 부분을 없앴다.

ehgud0670 committed:
[#step3] refactor: Beverages 파일 지우고 직접 값을 넣어 객체를 생성하도록 코드 수정
	이유: 일단 지우고 더 좋은 방안을 찾겠다.

ehgud0670 committed:
[#step3] refactor: Kind가 아닌 객체만으로 구별가능하도록 Hashable 채택함

ehgud0670 committed:
[#step3] refactor: 안에서 넣을 수 있는 프로퍼티는 생성자에서 파라미터 삭제

ehgud0670 committed:
[#step3] refactor: 리팩토링된 코드에 따라 기존 테스트 코드 수정

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverages/Beverage.swift:
```diff
@@ -8,48 +8,42 @@
     
+    static func == (lhs: Beverage, rhs: Beverage) -> Bool {
+        return lhs.hashValue == rhs.hashValue
+    }
```
hashable 사용한 방식이 좋네요.

ehgud0670 replied:
감사합니다 👍

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -8,41 +8,62 @@
+final class VendingMachine {
+    enum Notification {
+        static let beveragesDidChange = Foundation.Notification.Name("beveragesDidChange")
+        static let balanceDidChange = Foundation.Notification.Name("balanceDidChange")
```
이런 상수는 enum 타입보다는 struct가 어떨까요?
enum은 열거된 값을 사용하기 위해서 쓰는 건데 전혀 의미가 없을 것 같아요.

ehgud0670 replied:
넵, 저는 `struct` 대신 `enum`을 쓴 이유가 생성자가 제공되지 않는 자료형을 사용하기 위해서였습니다. 
enum은 열거된 값을 사용하기 위해 쓰는 것이 맞지만, [StyleShare-swift-style-guide] 에서 위의 이유로 enum을 사용하고, 저도 괜찮은것 같아 struct 대신 enum을 사용했습니다!

[StyleShare-swift-style-guide]: https://github.com/StyleShare/swift-style-guide

godrm replied:
그렇군요. 생성자가 제공되지 않으면 뭐가 좋은가요? :)

ehgud0670 replied:
음, struct나 class면 실수로 해당 객체를 생성할 수 있는데, enum이면 실수로 생성하더라도 생성할 수 없다고 컴파일 에러가 발생하므로 실수를 미연에 방지할 수 있을 것 같습니다.

godrm replied:
좋네요 :)

godrm commented on VendingMachineApp/VendingMachineApp/Models/Cashier.swift:
```diff
@@ -8,12 +8,20 @@
+    func isEnoughToBuy(price: Int) -> Bool
+    mutating func addToSalesLog(beverage: Beverage)
+    func currentBalance() -> Int
+    func searchSalesLog(handler: (Beverage) -> (Void))
```
여기 프로토콜 메소드들도 응집도가 높은지 확인해볼 필요가 있는 것 같습니다.
금액과 관련된 것과 음료 관련된 것이 섞여 있는 느낌입니다.

ehgud0670 replied:
넵, Stock 객체를 따로 만들어 음료 관련된 역할을 하도록 하고 Money 객체는 잔고와 관련된 역할만 하도록 다시 리팩토링하였습니다. 감사합니다

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -8,41 +8,62 @@
         cashier.plus(money: insertedMoney)
     }
     
     func currentMoney() -> Int {
```
이제는 이런 Money 자체도 Int로 표현하지 않고 Money 라는 타입을 만드는 게 가능한지, 필요한지, 필요하다면 어떤 기능이 필요한지 고려해보세요. Int, Double, String, Array 어떤 타입이든지 더 구체적인 타입을 만드는 걸 항상 고려하세요.

ehgud0670 replied:
넵 Money 객체를 만들어 객체끼리 비교 가능 하거나 객체 내부에서 돈을 빼고 추가할 수 있도록 하였습니다. 또 CustomStringConvertible 을 채택해서 바로 Int값인 돈을 반환하는 것보다 desciption을 사용하도록 하였습니다. 이렇게 함으로써 객체 내부로 코드를 옮기게 될 수 있었습니다. 감사합니다!

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -87,54 +106,39 @@ extension VendingMachine {
+                    return
             }
-            stockByKind[typeToString] = [beverage]
+            stockByKind[$0] = 1
```
guard else의 목적은 early exit가 크다고 생각합니다. 
이 경우는 그냥 있냐없냐에 따라 다른 로직을 처리하는 것 같아요. 그렇다면 굳이 부정 비교를 하면서까지 guard else로 표기하지 않아도 된다고 생각합니다.
부정 비교를 하면 그만큼 읽기 어렵습니다. 그냥 if 구문으로 해도 전혀 차이가 없는 것 같습니다. 
혹은 `() ? A : B` 삼항 연산자도 충분할 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -87,54 +106,39 @@ extension VendingMachine {
+                    return
             }
-            guard cashier.isEnoughToBuy(price: beverage.price)
+            guard cashier.isEnoughToBuy(price: $0.price)
```
여기 구문을 읽으니까 cashier 객체가 있는 게 오히려 혼동을 주는 것 같습니다.
캐셔가 돈이 부족하면 음료를 구매하지 못하는 건가요? 
반대로 사용자가 기계에 넣은 돈이 부족한 걸까요? 
객체로 타입을 만드는 것은 비유적인 표현이기도 합니다. 캐셔가 있는 게 저에게는 비유적으로 오해를 불러일으키네요. 어떤가요?

ehgud0670 replied:
넵, Cashier 자체가 말씀하신대로 매우 혼동을 주는 것 같습니다. ㅋㅋ 따라서 Cashier를 제거하고 , Money 객체를 따로 만들어 오해가 없게끔 리팩토링하였습니다. 감사합니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -8,29 +8,86 @@
+                let beverageNumber = stockByKind[beverage] {
+                $0.update(number: beverageNumber)
+            }
+        }
```
여기서 vendingMachine.stockByKind() 로 값을 가져와서 몇 개 있는지 찾기보다는 
`vendingMachine.stockBy(kind: beverage)` 형태로 요청하는 게 어떨까요?

ehgud0670 replied:
넵, 수정하였습니다. 훨씬 사용할 목적에 맞는 메소드인 것 같습니다. 감사합니다

godrm commented on VendingMachineApp/VendingMachineApp/Views/BalanceButtons/Balance5000Button.swift:
```diff
@@ -0,0 +1,15 @@
+
+final class Balance5000Button: BalanceButton {
+    override func invokeAction(sender: BalanceButton) {
+        action?(5000)
```
1000, 5000 값도 속성으로 표현하면 더 간결하게 표현할 수 있지 않을까요?

ehgud0670 replied:
넵, 속성으로 표현하였습니다. 감사합니다.

godrm commented on VendingMachineApp/VendingMachineApp/Views/BeverageButtons/CantataPrimiumLatte175Button.swift:
```diff
@@ -0,0 +1,22 @@
+            celsius: 65,
+            name: "프리미엄 라떼",
+            volume: 175,
+            price: 1500
```
뷰에 객체를 만드는 코드가 하드코딩 된 것은 좀 아쉽습니다. 버튼의 액션을 추상화하는 것은 이런 코드를 직접 구현하지 않고, 다른 객체에서 처리하도록 하기 위한 겁니다. 
뷰와 모델을 분리하고, 모델 객체를 생성하는 것도 분리해보세요.

ehgud0670 replied:
넵, 그래서 고민하다가 음료를 생성하는 BeverageInstantiator를 만들어 연산 프로퍼티로  각각의 객체를 리턴하도록 하였습니다.
객체지향적으로 상속을 고민했지만, 오히려 이렇게 생성만 하는 객체는 절차지향적으로 만드는 것이 코드를 이해하는게 더 빠를 것 같아서 이렇게 작성했습니다.

ehgud0670 replied:
아직 고민중입니다

ehgud0670 replied:
넵, 감사합니다. 뷰와 모델 객체를 분리하기 위해서 모델을 생성하는 객체를 구현하였습니다. 👍

godrm reviewed (commented):
리뷰하다가 완료 처리가 늦었네요. 다른 관점에서 의견을 남겨봤습니다.

ehgud0670 committed:
[#step3] feat: guard 문을 if문으로 대체함, 테스트로 해당 메소드가 체크함
	guard 문을 else 문으로 바꾼 이유:  guard else 문은 early exit, 즉 조건이 안되면 거르는 용도로 사용해야 한다.
					   따라서  else 문에 에러를 날리든지 해야지 로직이 들어가면 안된다.
					   근데 이전의 코드는 guard else 문에 로직이 들어가 있으므로 이를 if문으로 바꾼다.

ehgud0670 committed:
[#step3] refactor: balance를 속성으로 추가하고 메소드에 적용

ehgud0670 committed:
[#step3] refactor: 음료를 관리하는 Stock 객체를 만들어 응집성을 높임
		 이유: Cashier의 역할은 돈을 관리하는 것인데 음료 관련 메소드가 있으므로 이는 단일 책임 원칙에 위배한다.
			이를 발견하고 Beverage 를 책임지는 Stock 객체를 따로 만들어 응집성을 높인다.

ehgud0670 committed:
[#step3] refactor: 노티를 보낼때 음료, 잔고를 보내도록 코드 수정
 	이유: 음료를 노티로 보내면 해당 음료만 업데이트 하면 되므로 시간복잡도를 줄일 수 있다.

ehgud0670 committed:
[#step3] refactor: calculable 도 의존성 주입하도록 기능 수정

ehgud0670 committed:
[#step3] style: search 보단 repeat 이 맞는것 같아 메소드들 이름 수정

ehgud0670 committed:
[#step3] refactor: 아직 유통기한이 지났는지 검증이 안된 객체이므로 milk로 수정

ehgud0670 committed:
[#step3] refactor: Cashier 에서 Money로 이름 변경하고 더 커밋 메시지 작성하기 ^^

ehgud0670 committed:
[#step3] refactor: 해당 음료의 개수가 몇개인지 Stock 내부에서 찾도록 코드 수정

ehgud0670 committed:
[#step3] refactor: 구체타입인 Int가 아닌 Int를 담는 Money 객체로 변경

ehgud0670 committed:
[#step3] refactor: 음료 모델을 생성하는 객체를 만들어 절차지향적으로 여러 음료를 반환하도록 함

ehgud0670 committed:
[#step3] refactor: 반환하는 음료 타입을 음료의 sub-class 로 변경

ehgud0670 committed:
[#step3] refactor: 뷰 객체가 음료 생성 객체를 반환하고 음료 생성 객체가 음료를 반환하는 구조로 수정
		* 이를 위해 음료 생성 객체도 상속하는 구조로 변경

ehgud0670 committed:
[#step3] refactor: 뷰가 머니 생성 객체를 반환하고 머니 생성 객체가 머니를 반환하는 구조로 변경

	이유: 뷰와 모델을 분리하기 위해 모델을 생성하는 객체를 만들었고 이를 적용하였다.

ehgud0670 commented:
모든 피드백을 완료하였습니다

ehgud0670 committed:
[#step3] docs: 바뀐 UI에 따라 이미지 업데이트
